### PR TITLE
implement custom mouse function framework

### PR DESCRIPTION
**Pull request by kne42**
# Description

For on press, on drag, and on release events:
```python
@layer.mouse_drag_callbacks.append
def hello_world(layer, event):
    "dragging"
    # on press
    print('hello world!')
    yield

    # on move
    while event.type == 'mouse_move':
        print(event.pos)
        yield

    # on release
    print('goodbye world ;(')
```

For only an on press event:
```python
@layer.mouse_drag_callbacks.append
def press(layer, event):
    print('hi')
```

For whenever the mouse is moved - no pressing needed:
```python
@layer.mouse_move_callbacks.append
def move(layer, event):
    print(event.pos)
```

double click is not supported

### CODE REVIEW

**Comment by sofroniewn** [comment]
I've added an example of a custom mouse function that does a dilation on the clicked region in a labels layer and prints to the status bar the size of the region after dilation. It's in `examples/custom_mouse_functions.py`. Here's a gif:

custom_mouse_functions

Pretty cool stuff I think! We still have to think about the API / naming a little, and I still need to add tests but thanks to Kira this is coming along very nicely!!

On the API side, we need to think about if it is possible to bind custom mouse functions to layer classes, not just instances, and if it is possible to bind to the viewer and the Viewer class.

Right now we've made the mouse functions be additive which I think is the right choice.

I'm not 100% sure about the `@layer.mouse_drag_callbacks.append` syntax, and if it's clear that that's where mouse press / release get added too, and if it's clear what the difference with `@layer.mouse_move_callbacks.append` is, but we can discuss that next meeting.

Maybe @royerloic @jni @AhmetCanSolak have thoughts

**Comment by sofroniewn** [comment]
Note that this will being to address #252 and #531 from @VolkerH

**Comment by sofroniewn** [comment]
This is now ready for review @jni. I've added what I hope are pretty comprehensive tests, including simulating mouse press, drag, release, and move using `from unittest.mock import Mock`. Thanks for the tip @shanaxel42! (I will probably make a new PR adding similar tests for our keybindings if people like them, and other aspects of the code - `Mock` seems like it will be very useful for testing a lot of the UI interactivity stuff)

I've made it so that a mouse binding added to `mouse_move_callbacks` will NOT get called during a drag event even though the mouse is technically moving. How do people feel about that? I thought it was cleaner to have it completely separate from anything that might get called in `mouse_drag_callbacks`.

I've also made it so you can now bind custom mouse functions to the `viewer` not just the layers. I havn't made them bindable to the class `Viewer`, but to the instance, but I think that's ok. We might even want to remove the class bindings from the keys at some point but that can wait.

I've also update the example now so that clicking will cause a dilation but alt clicking will cause an erosion to illustrate how you can use modifier keys inside custom mouse functions.

Finally if people like this, then I will move the custom mouse functions inside all the layers to a separate file called `mousebindings.py` that will look like the `keybindings.py` file in some senses. I'll do that in a new PR though. That will make it easier to overwrite any mousebindings that we've added instead of it just being additive.

**Comment by sofroniewn** [comment]
So the I'm pretty confident the reason why our windows build is failing in this PR and this PR alone is I added the line `self.canvas.events.ignore_callback_errors = False`. This wasn't something needed for this PR, but is probably good practice to have as otherwise there could be errors that we aren't catching. I think we should make a new PR adding that line in, and my suspicion is that the windows tests that are failing could be related to the windows bugs that people have been reporting in #527 and and #428 but which we were missing.

File: examples/custom_mouse_functions.py
Comment (jni) [inline examples/custom_mouse_functions.py]:
```suggestion
            msg = f'clicked at {cords} on blob {val} which is now {size} pixels'
```

File: examples/custom_mouse_functions.py
Comment (jni) [inline examples/custom_mouse_functions.py]:
```suggestion
            msg = f'clicked at {cords} on background ({val}) which is ignored'
```

File: napari/_qt/qt_viewer.py
Comment (jni) [inline napari/_qt/qt_viewer.py:304]: Why is this needed?

File: napari/_qt/qt_viewer.py
Comment (jni) [inline napari/_qt/qt_viewer.py]: All this stuff is very confusing... I think comments would be very welcome in code like this, explaining what's happening.

File: napari/_qt/qt_viewer.py
Comment (jni) [inline napari/_qt/qt_viewer.py]: Should these be in a `finally`? ie do they need to be activated whether or not an uncaught exception happens in the try clause?

File: napari/_qt/qt_viewer.py
Comment (jni) [inline napari/_qt/qt_viewer.py]: This code is repeated in all the `on_mouse_*` methods... Would it be possible to unify it somewhere?

**Review by jni — APPROVED** [review]
@sofroniewn I don't have time for an in depth review right now. tbh I would prefer if you could hold off on merging this, so that we could get it to a simpler, more easily grokkable state, but if you want it for your talk, go for broke. ;)

**Comment by royerloic** [comment]
This is very cool!
Happy with how it is implemented now,
usage will dictate if changes are needed.
 
I think the dilation/erosion feature is super useful, 
and could become a 'tool' in the label layer, perhaps.

**Comment by sofroniewn** [comment]
@jni @royerloic I still need to make these changes - will do so now

File: napari/_qt/qt_viewer.py
Comment (sofroniewn) [reply napari/_qt/qt_viewer.py:304]: I asked kira about this, and she said it is needed to make sure the custom function don't edit the events. While this seems unlikely I guess it is safer.

File: examples/custom_mouse_functions.py
Comment (sofroniewn) [reply examples/custom_mouse_functions.py]: fixed

File: examples/custom_mouse_functions.py
Comment (sofroniewn) [reply examples/custom_mouse_functions.py]: fixed

File: napari/_qt/qt_viewer.py
Comment (sofroniewn) [reply napari/_qt/qt_viewer.py]: possibly, i'd be very open to suggestions. maybe can come in a refactor

File: napari/_qt/qt_viewer.py
Comment (sofroniewn) [reply napari/_qt/qt_viewer.py]: i'm not 100% sure

Comment (jni) [reply napari/_qt/qt_viewer.py]: I find it scary to merge code that we don't understand fully... There are parts of scikit-image that were merged with only a superficial understanding by the reviewer, and this code has ended up being very hard to maintain — essentially untouchable. So if you can live without this in 0.2 I'd love to wait for us to be able to review this in detail and understand deeply what's happening. (But if you can't... Well, we'll deal with it. ;)

**Comment by sofroniewn** [comment]
Ok, I'm going to wait on this for `0.2.0`, but would like to add it to `0.2.1` so if we could do a through job reviewing this relatively soon @jni that would be great. I can look at it in more detail again next week.

File: napari/tests/test_mouse_bindings.py
Comment (sofroniewn) [inline napari/tests/test_mouse_bindings.py:2]: @jni you mentioned using `pytest` mock, but there doesn't seem to be one built in - it looks like a separate package from here - https://github.com/pytest-dev/pytest-mock - but maybe I'm getting this wrong. Otherwise I think leaving the `unittest` one is fine

File: napari/_qt/qt_viewer.py
Comment (sofroniewn) [reply napari/_qt/qt_viewer.py]: unified in `utils.misc`

File: napari/_qt/qt_viewer.py
Comment (sofroniewn) [reply napari/_qt/qt_viewer.py]: i think they are in the right place as these objects need to be deleted at the end of the mouse release regardless of if there was an error or not

**Comment by sofroniewn** [comment]
@jni I have refactored this to reduce duplicate code and added a couple comments. I'm pretty happy with it now and think we should merge, I leave it to you to do a deep dive though. It would be nice to get it in soon though

File: napari/components/viewer_model.py
Comment (jni) [inline napari/components/viewer_model.py:102]: Can you add comments for each of these elements about what they are supposed to hold? It's not totally obvious from the name.

File: napari/tests/test_mouse_bindings.py
Comment (jni) [reply napari/tests/test_mouse_bindings.py:2]: Probably the package is what I was thinking about. Happy to stay with unittest for now since we are not using complex features within it (that might require less pytest-friendly syntax), and it's built in.

File: napari/util/misc.py
Comment (jni) [inline napari/util/misc.py:599]: So imho all this logic needs comments explaining what is being done, though I sort of get it:

- run the function, getting a result
- if the result is a generator, then get the next value from the generator, and
- ??? What is being added to those dictionaries? And where are the dictionaries used?

In general, one of the principles of writing bug free code is that code should be as local as possible, meaning that I shouldn't have to keep in my head all the possible accesses and lifetimes of stuff that lives outside of this function. I'm not sure how to improve this implementation but at a minimum, there should be comments here about what those dictionaries do and where they are "cleared" and why. (I see some dels in the next function but it's unclear to me whether they refer to the same object (ie are they supposed to happen in quick succession)

**Review by jni — COMMENTED** [review]
@sofroniewn I sort of get the implementation now, but I think it needs some comments around those dictionary objects to make it clear to future generations what's happening where. At that point is should be mergeable.

File: napari/components/viewer_model.py
Comment (sofroniewn) [reply napari/components/viewer_model.py:102]: added

File: napari/util/misc.py
Comment (sofroniewn) [reply napari/util/misc.py:599]: done

**Comment by sofroniewn** [comment]
added more comments and the example code for defining one of these callbacks to the doc strings - i feel pretty good about this now

I will merge once the tests pass

**Pull request merged**
